proposals/0239-operator-designated-types.md

```diff
@@ -0,0 +1,137 @@
+
+The goal is to use this hint to replace the currently undocumented (and difficult to reason about) hacks in the type checker that are in place today and which are required in order to typecheck even relatively simple expressions.
+
+## Detailed design
+
+Operator declaration syntax is extended to allow for one or more designated types to be specified. For infix operators, these types are listed after the (optional) precedence group. For prefix and postfix operators, these types come immediately after a `:` following the operator name. Some examples:
+
+- `infix operator   * : MultiplicationPrecedence, Numeric`
+- `infix operator   / : MultiplicationPrecedence, BinaryInteger, FloatingPoint`
+- `prefix operator ~ : BinaryInteger`
+
+Specifying one or more types results in the expression type checker looking at declarations defined in those types before other declarations of the operator, and stopping as soon as it finds that one of the declarations in one of the types can be used to successfully type check the expression. If it fails to typecheck successfully with one of those declarations, it continues attempting all the other overloads for that operator.
+
+In cases where there are multiple designated types for a given operator, the specific order in which the overloads in those types are examined is left unspecified, but should produce results that are compatible with the existing type checker's notion of the "best" solution if attempting all of the overloads could result in multiple solutions.
+
+## Experimental results
+
+The implementation for this is already in-tree, and is enabled as such:
+
+- `-Xfrontend -enable-operator-designated-types`: enables parsing the extended operator declarations
+- `-Xfrontend -solver-enable-operator-designated-types`: enables having the constraint solver make use of the declarations
+
+The current implementation has been tested across 25 test cases that were known to be slow and/or result in the type checker giving up after expending substantial effort attempting to type check the expression.
+
+In addition to enabling this new *designated type* functionality, this testing disables the `shrink` portion of the expression type checker (via `-Xfrontend -solver-disable-shrink`).  This portion of the expression type checker is known to help occassionally but often results in substantially increasing the expression type checking time since it frequently doesn't provide any benefit but does a lot of work. We also disable a number of other problematic hacks (via `-Xfrontend -disable-constraint-solver-performance-hacks`).
```

Comment:
occassionally => occasionally

```suggestion
In addition to enabling this new *designated type* functionality, this testing disables the `shrink` portion of the expression type checker (via `-Xfrontend -solver-disable-shrink`).  This portion of the expression type checker is known to help occasionally but often results in substantially increasing the expression type checking time since it frequently doesn't provide any benefit but does a lot of work. We also disable a number of other problematic hacks (via `-Xfrontend -disable-constraint-solver-performance-hacks`).
```

Reply:
Fixed in e054585.